numpy/random/mtrand/mtrand.pyx

```diff
@@ -1142,7 +1142,9 @@ cdef class RandomState:
                 raise ValueError("a and p must have same size")
             if np.logical_or.reduce(p < 0):
                 raise ValueError("probabilities are not non-negative")
-            if abs(kahan_sum(pix, d) - 1.) > atol:
+
+            # negated to handle NaNs in p
```

Comment:
Thinking about it, this would make sense to do to the `p < 0` above instead - that would make the warning dissappear to.

So `if not np.all(p >= 0): raise ...`